google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,406 @@
+# Copyright 2017 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Rules engine for Bucket retention."""
+import collections
+from collections import namedtuple
+import itertools
+import threading
+import json
+
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.util import date_time as dt
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.common.util import relationship
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+
+
+LOGGER = logger.get_logger(__name__)
+
+SUPPORTED_RETENTION_RESOURCE_TYPES = frozenset(['bucket'])
+
+VIOLATION_TYPE = 'RETENTION_VIOLATION'
+_APPLY_TO_BUCKETS = 'bucket'
```

Comment:
You can get rid of `_APPLY_TO_BUCKETS` and `_APPLY_TO_RESOURCES` and just use `SUPPORTED_RETENTION_RESOURCE_TYPES` instead.